modules/@apostrophecms/translation/index.js

```diff
@@ -1,17 +1,173 @@
+    self.providers = [];
 
     self.enableBrowserData();
   },
+  handlers(self) {
+    return {
+      '@apostrophecms/doc-type:beforeLocalize': {
+        // Translate the document using the first available provider.
+        async translate(req, doc, source, target, options) {
+          if (!req.query.aposTranslateTargets?.includes(target)) {
+            return;
+          }
+          // We don't support multiple providers yet, so just use the first one
+          // when no provider is specified.
+          const name = self.getProvider(req.query.aposTranslateProvider)?.name;
+          const module = self.getProviderModule(name);
+          if (!name || !module) {
+            const name = req.query.aposTranslateProvider ||
+              'apostrophe:notAvailable';
+
+            self.logError('before-localize-translate', 'Provider not found.', {
+              _id: doc._id,
+              title: doc.title,
+              provider: name
+            });
+            return self.apos.notify(
+              req,
+              req.t(
+                'apostrophe:automaticTranslationErrorNoProvider',
+                {
+                  title: doc.title,
+                  name: req.t(name)
+                }
+              ),
+              {
+                type: 'danger',
+                dismiss: true
+              }
+            );
+          }
+
+          return module.translate(req, name, doc, source, target, options);
+        }
+      }
+    };
+  },
+  apiRoutes: (self) => ({
+    get: {
+      // Get the list of supported languages for the given provider.
+      // If no provider is specified, the first available provider is used.
+      // query parameters:
+      // - provider: The name of the provider
+      // - source: Array of language codes to translate from
+      // - target: Array of language codes to translate to
+      // Example response:
+      // ```json
+      // {
+      //   source: [
+      //     { code: 'en', supported: true },
+      //     { code: 'de', supported: true }
+      //   ],
+      //   target: [
+      //     { code: 'en', supported: true },
+      //     { code: 'de', supported: true }
+      //   ]
+      // }
+      // ```
+      // If no `source` or `target` parameters are provided, the response will
+      // contain all supported languages for any source or target language.
+      async languages(req) {
+        if (!self.canAccessApi(req)) {
+          throw self.apos.error('notfound');
+        }
+
+        const name = self.getProvider(req.query.provider)?.name;
+        const module = self.getProviderModule(name);
+        if (!name || !module) {
+          throw self.apos.error(
+            'invalid',
+            req.t('apostrophe:automaticTranslationLngCheckNoProvider')
+          );
+        }
+
+        return module.getSupportedLanguages(
+          req,
+          name,
+          req.query.source ?? [],
+          req.query.target ?? []
+        );
```

Comment:
[optional]
For readability I would pass req as a first argument and the rest in  an object
```javascript
      return module.getSupportedLanguages(
          req,
          {
            name,
            source: req.query.source ?? [],
            target: req.query.target ?? []
          }
        );
```

Reply:
Also this one should be "laundered" I guess.